community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/InputGroupsDeserializerTest.java

```diff
@@ -80,7 +84,52 @@ public void shouldBeAbleToAskForSourceInformationEvenBetweenTwoSources() throws
         assertEquals( 2, flips.get() );
     }
 
-    private Configuration lowBufferSize( Configuration conf )
+    @Test
+    public void shouldCoordinateGroupCreationForParallelProcessing() throws Exception
+    {
+        // GIVEN
+        List<DataFactory<InputNode>> data = new ArrayList<>();
+        int processors = Runtime.getRuntime().availableProcessors();
+        for ( int i = 0; i < processors; i++ )
+        {
+            StringBuilder builder = new StringBuilder( ":ID(Group" + i + ")" );
+            for ( int j = 0; j < 100; j++ )
+            {
+                builder.append( "\n" + j );
+            }
+            data.add( data( builder.toString() ) );
+        }
+        Groups groups = new Groups();
+        IdType idType = IdType.INTEGER;
+        Collector badCollector = mock( Collector.class );
+        Configuration config = lowBufferSize( COMMAS, false );
+        try ( InputGroupsDeserializer<InputNode> deserializer = new InputGroupsDeserializer<>(
+                data.iterator(), defaultFormatNodeFileHeader(), config, idType,
+                processors, processors, (header,stream,decorator,validator) ->
+                {
+                    InputNodeDeserialization deserialization =
+                            new InputNodeDeserialization( header, stream, groups, idType.idsAreExternal() );
+                    return new InputEntityDeserializer<>( header, stream, config.delimiter(),
+                            deserialization, decorator, validator, badCollector );
```

Comment:
This is the same lamba used in `CsvInput#nodes()` and almost the same as `CsvInput#relationships()`.
I think readability would be improved by extracting them to concrete classes.